core/src/main/java/hudson/diagnosis/OldDataMonitor.java

```diff
@@ -163,15 +157,19 @@ public void onDeleted(Run run) {
      */
     public static void report(Saveable obj, String version) {
         OldDataMonitor odm = get(Jenkins.getInstance());
-        synchronized (odm) {
-            try {
-                SaveableReference ref = referTo(obj);
+        try {
+            SaveableReference ref = referTo(obj);
+            while (true) {
                 VersionRange vr = odm.data.get(ref);
-                if (vr != null) vr.add(version);
-                else            odm.data.put(ref, new VersionRange(version, null));
-            } catch (IllegalArgumentException ex) {
-                LOGGER.log(Level.WARNING, "Bad parameter given to OldDataMonitor", ex);
+                if (vr != null) {
+                    vr.add(version);
+                    break;
+                } else if (odm.data.putIfAbsent(ref, new VersionRange(version, null)) == null) {
```

Comment:
Minor performance :ant: : Seems that non-null putIfAbsent() result leads to the VersionRange retrieval on the next iteration => we effectively request it two times on the collision.